src/core/abr/learn_2_adapt.ts

```diff
@@ -0,0 +1,260 @@
+  prevW: number[];
+  Q: number;
+  segmentRequestStart: number;
+  segmentDownloadFinish: number;
+  bufferTargetLvl: number;
+}
+
+interface ICxtArg {
+  bitrates: number[];
+  representations: Representation[];
+  streamEvents$: Observable<IABRStreamEvents>;
+}
+
+class L2ARule {
+  // private readonly _representations: Representation[];
+  private readonly _bitrates: number[];
+
+  private _state: L2A_State;
+  private _lastQuality: number;
+  // private _placeholderBuffer: number;
+  // private _lastSegmentWasReplacement: boolean;
+  private _mostAdvancedSegmentStart: number | null;
+  private _lastSegmentDuration: number | null;
+  private _lastSegmentStart: number | null;
+  private _lastSegmentRequestTimeMs: number | null;
+  private _lastSegmentFinishTimeMs: number | null;
+
+  private _l2AParameters: L2AParameters;
+
+  constructor({ bitrates }: ICxtArg) {
+    // this._representations = representations;
+    this._bitrates = bitrates;
+
+    this._state = L2A_State.Start;
+    this._lastQuality = 0;
+    this._lastSegmentDuration = 0;
+    // this._placeholderBuffer = 0;
+    // this._lastSegmentWasReplacement = false;
+    this._mostAdvancedSegmentStart = null;
+    this._lastSegmentDuration = null;
+    this._lastSegmentStart = null;
+    this._lastSegmentRequestTimeMs = null;
+    this._lastSegmentFinishTimeMs = null;
+
+    this._l2AParameters = {
+      w: [],
+      prevW: [],
+      Q: 0,
+      segmentRequestStart: 0,
+      segmentDownloadFinish: 0,
+      bufferTargetLvl: 1.5,
+    };
+  }
+
+  private _resetL2AInfos() {
+    // this._placeholderBuffer = 0;
+    // this._lastSegmentWasReplacement = false; // used to know if the last segment has been replace (we dont use it for now);
+    this._mostAdvancedSegmentStart = null;
+    this._lastSegmentDuration = null;
+    this._lastSegmentStart = null;
+    this._lastSegmentRequestTimeMs = null;
+    this._lastSegmentFinishTimeMs = null;
+  }
+
+  /**
+   * Adjust State accordingly
+   */
+  private _checkNewSegment() {
+    if (this._lastSegmentStart !== null &&
+        this._lastSegmentRequestTimeMs !== null &&
+        this._lastSegmentFinishTimeMs !== null) {
+      this._l2AParameters.segmentRequestStart = 0.001 * this._lastSegmentRequestTimeMs;
+      this._l2AParameters.segmentDownloadFinish = 0.001 * this._lastSegmentFinishTimeMs;
+      this._lastSegmentStart = null;
+      this._lastSegmentRequestTimeMs = null;
+    }
+  }
+
+  private _dotmultiplication(a: number[], b: number[]) {
+    if (a.length !== b.length) {
+      return -1;
+    }
+    let sumdot = 0;
+    for (let i = 0; i < a.length; i++) {
+      sumdot = sumdot + a[i] * b[i];
+    }
+    return sumdot;
+  }
+
+  /**
+     * Project an n-dim vector y to the simplex Dn
+     * Dn = { x : x n-dim, 1 >= x >= 0, sum(x) = 1}
+     * Algorithm is explained at http://arxiv.org/abs/1101.6081
+     * @param {array} arr
+     * @return {array}
+     */
+  private _euclideanProjection(arr: number[]): number[] {
```

Comment:
It doesn't seem to rely on this class's state at all. It might make more sense as its separate function